### DISCUSSION TITLE
Triton effect can not display in osgEarth3.6

### DISCUSSION THREAD

**Discussion by FuXiii** [pelicanmapping/osgearth#2600] [Support Q&A]
* GPU: `RTX 3070`

Before I use `osgEarth3.6` , I had been used `osgEarth3.3` : install the `osgEarth::Util::LogarithmicDepthBuffer` on `root` camera and use `user-functions.glsl` `user-vert-functions.glsl` which under the `src/osgEarthTriton/Shaders` to replace the same name shader file under `$(TritonPath)/Resources`

Everything is well in `osgEarth3.3` , but after I upgrade to use `osgEarth3.6`  the `Triton`  ocean disappear base the same code usage.

And I try to:
1.  uninstall `osgEarth::Util::LogarithmicDepthBuffer`
2.  change the `user-vert-functions.glsl` like:
```glsl
...
vec4 overridePosition(in vec4 position)
{
   //return oe_logdepth_vert(position); // osgEarth3.6
   return position;
}
```
3. change `user-functions.glsl` like:
```glsl
...
void writeFragmentData(in vec4 finalColor, in vec4 Cdiffuse, in vec3 lightColor, in vec3 nNorm)
{
   //finalColor.a *= oe_VisibleLayer_opacityUniform;// osgEarth3.6
   finalColor.a = 1;

#ifdef OPENGL32
   fragColor = finalColor;
#else
   gl_FragColor = finalColor;
#endif
}
```
The `Triton`  ocean can display but looks like the `accuracy` is low (Not highly coupled with the coastline).

After that I guess maybe the `uniform mat4 osg_ProjectionMatrix` not set correctly, and then I try to code in `user-vert-functions.glsl` like:
```glsl
...
vec4 overridePosition(in vec4 position)
{
   mat clip2view = inverse(osg_ProjectionMatrix);
   return osg_ProjectionMatrix * clip2view * position;
}
```

And after I run the application the `Triton` ocean disappear. Maybe the `osg_ProjectionMatrix` not set correctly let it become `zero` matrix.

---

And I try to use `TritonLayer::setOpacity(float)` to set ocean `Alpha` , the result look like not affect on `oe_VisibleLayer_opacityUniform` in `user-functions.glsl`

**Comment by gwaldron** [comment]
Thanks for explaining the things you tried.
My Triton is working properly with 3.6 with the stock changes. Did you update Triton to the latest version as well? What version are you using? I am on version 4.28.

**Reply by FuXiii** [reply]
base `releasenotes.txt` under `Triton SDK` :

```text
Triton Release History

Version 4.50 March 22, 2023
- Rebuilt fftss library on Windows with vc14 to avoid linker warnings
- Fix a couple of compiler warnings

...
```

the version is `4.50`

**Reply by gwaldron** [reply]
Correction, I am also using 4.50. 
(I was looking in the `Version.h` file which said 4.28)

**Reply by FuXiii** [reply]
Base my install  `Triton`, there has not `Version.h` file.
![image](https://github.com/user-attachments/assets/a70465a4-a189-4d7a-ad73-11fc6c925742)

So when I compile `osgEarth` I had remove the `#include<Version.h>` in:
* SliverLiningAPIWapper.cpp
* SliverLiningCloudDrawable.cpp
* TritonDrawable.cpp

*note : The `Triton` install package download from [Sundog](https://www.sundog-soft.com/features/ocean-and-water-rendering-with-triton/)*

**Reply by gwaldron** [reply]
I just downloaded the latest version (4.54) and there`Version.h` is there.
Built and tested OK after copying in the shaders.
